components/rendering/src/markdown.rs

```diff
@@ -1,20 +1,121 @@
 use std::borrow::Cow::Owned;
+use std::iter::Peekable;
 
 use pulldown_cmark as cmark;
 use self::cmark::{Parser, Event, Tag, Options, OPTION_ENABLE_TABLES, OPTION_ENABLE_FOOTNOTES};
 use slug::slugify;
 use syntect::easy::HighlightLines;
 use syntect::html::{start_coloured_html_snippet, styles_to_coloured_html, IncludeBackground};
 
-use errors::Result;
+use errors::{Result, Error};
 use utils::site::resolve_internal_link;
 use context::Context;
 use highlighting::{SYNTAX_SET, THEME_SET};
 use short_code::{SHORTCODE_RE, ShortCode, parse_shortcode, render_simple_shortcode};
 use table_of_contents::{TempHeader, Header, make_table_of_contents};
 
+struct GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+    context: &'a Context<'a>,
+    errors: &'c mut Vec<Error>,
+    parser: Peekable<Parser<'b>>,
+}
+
+fn is_internal_link(link: &str) -> bool {
+    link.starts_with("./")
+}
+
+impl<'a, 'b, 'c> GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+    fn new(context: &'a Context<'a>, parser: Parser<'b>, errors: &'c mut Vec<Error>) -> GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+        GutenbergFlavoredMarkdownParser {
+            context: context,
+            errors: errors,
+            parser: parser.peekable()
+        }
+    }
+
+    fn push_error<E: Into<Error>>(&mut self, error: E) {
+        self.errors.push(error.into());
+    }
+
+    fn empty_event(&self) -> Event<'b> {
+        Event::Html(Owned("".to_string()))
+    }
+
+    fn process_link(&mut self, link: Tag<'b>) -> Event<'b> {
+        match link {
+            Tag::Link(ref href, ref text) => {
+                if is_internal_link(href) {
+                    match resolve_internal_link(href, self.context.permalinks) {
+                        Ok(url) => {
+                            Event::Start(Tag::Link(Owned(url), text.clone()))
+                        },
+                        Err(_) => {
+                            self.push_error(format!("Relative link {} not found.", href));
+                            self.empty_event()
+                        }
+                    }
+                } else{
+                    Event::Start(Tag::Link(href.clone(), text.clone()))
```

Comment:
not possible to return the `link` without cloning?

Reply:
The current implementation clones as well, so we just left it while it's still a WIP, but we'll go back at the end and clean up as many clones as possible.